apps/web/src/components/providers/EnvironmentProvider.tsx

```diff
@@ -0,0 +1,178 @@
+
+function selectEnvironment(environments: IEnvironment[] | undefined, selectedEnvironmentId?: string) {
+  let e: IEnvironment | undefined;
+
+  if (!environments) {
+    return;
+  }
+
+  // Find the environment based on the current user's last environment
+  if (selectedEnvironmentId) {
+    e = environments.find((env) => env._id === selectedEnvironmentId);
+  }
+
+  // Or pick the development environment
+  if (!e) {
+    e = environments.find((env) => env.name === BaseEnvironmentEnum.DEVELOPMENT);
+  }
+
+  if (!e) {
+    throw new Error('Missing development environment');
+  }
+
+  saveEnvironmentId(e._id);
+
+  return e;
+}
+
+export function EnvironmentProvider({ children }: { children: React.ReactNode }) {
+  const navigate = useNavigate();
+  const queryClient = useQueryClient();
+  const { currentOrganization } = useAuth();
+  const {
+    data: environments,
+    isLoading,
+    refetch: refetchEnvironments,
+  } = useQuery<IEnvironment[]>([QueryKeys.myEnvironments, currentOrganization?._id], getEnvironments, {
+    enabled: !!currentOrganization,
+    retry: false,
+    staleTime: Infinity,
+  });
+
+  const [currentEnvironment, setCurrentEnvironment] = useState<IEnvironment | undefined>(
+    selectEnvironment(environments, getEnvironmentId())
+  );
+
+  const switchEnvironment = useCallback(
+    async ({ environmentId, redirectUrl }: Partial<{ environmentId: string; redirectUrl: string }> = {}) => {
+      setCurrentEnvironment(selectEnvironment(environments, environmentId));
+
+      /*
+       * TODO: Replace this revalidation by moving environment ID or name to the URL.
+       * This call creates an avalance of HTTP requests as the more you navigate across the app in a
+       * single environment the more invalidations will be triggered on environment switching.
+       */
+      await queryClient.invalidateQueries();
+
+      if (redirectUrl) {
+        await navigate(redirectUrl);
+      }
+    },
+    [queryClient, navigate, setCurrentEnvironment, environments]
+  );
+
+  const switchToProductionEnvironment = useCallback(
+    async (redirectUrl?: string) => {
+      const environmentId = environments?.find((env) => env.name === BaseEnvironmentEnum.PRODUCTION)?._id;
+
+      if (environmentId) {
+        await switchEnvironment({
+          environmentId,
+          redirectUrl,
+        });
+      } else {
+        throw new Error('Production environment not found');
+      }
+    },
+    [environments, switchEnvironment]
+  );
+
+  const switchToDevelopmentEnvironment = useCallback(
+    async (redirectUrl?: string) => {
+      const environmentId = environments?.find((env) => env.name === BaseEnvironmentEnum.DEVELOPMENT)?._id;
+
+      if (environmentId) {
+        await switchEnvironment({
+          environmentId,
+          redirectUrl,
+        });
+      } else {
+        throw new Error('Development environment not found');
+      }
+    },
+    [environments, switchEnvironment]
+  );
```

Comment:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
✏ **suggestion:** ‏I think we could generalize this, even if the intent is to have two separate functions for specific env switch, e.g. something like:

```typescript
const switchToEnvironment = useCallback(
  async (environmentName: BaseEnvironmentEnum, redirectUrl?: string) => {
    const environmentId = environments?.find((env) => env.name === environmentName)?._id;

    if (environmentId) {
      await switchEnvironment({
        environmentId,
        redirectUrl,
      });
    } else {
      throw new Error(`${capitalize(environmentName.toLowerCase())} environment not found`);
    }
  },
  [environments, switchEnvironment]
);


const switchToProductionEnvironment = (redirectUrl?: string) => switchToEnvironment(BaseEnvironmentEnum.PRODUCTION, redirectUrl);
const switchToDevelopmentEnvironment = (redirectUrl?: string) => switchToEnvironment(BaseEnvironmentEnum.DEVELOPMENT, redirectUrl);
```